### PR TITLE
chore: shows network state in ethers & wagmi lab pages

### PR DESCRIPTION
**Pull request by ganchoradkov**
# Description
Shows popups on wss connect/disconnect in ethers & wagmi lab pages to better monitor wss connection especially during ci runs

## Type of change

- [x] Chore (non-breaking change that addresses non-functional tasks, maintenance, or code quality improvements)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# Showcase (Optional)

image

# Checklist

- [ ] Code in this PR is covered by automated tests (Unit tests, E2E tests)
- [ ] My changes generate no new warnings
- [ ] I have reviewed my own code
- [ ] I have filled out all required sections

### CODE REVIEW

File: apps/laboratory/src/pages/library/ethers.tsx [hunk 1]
```diff
@@ -23,8 +26,36 @@ const modal = createWeb3Modal({
 })
 
 ThemeStore.setModal(modal)
-
+let loaded = false
 export default function Ethers() {
+  const toast = useChakraToast()
+
+  React.useEffect(() => {
```
[the remark this answers is not in the document]
Comment (ganchoradkov) [reply apps/laboratory/src/pages/library/ethers.tsx:33]: `useEffect`doesn't accept promises. We can potentially use `useMemo` which accepts promises but seems incorrect usage

File: apps/laboratory/src/pages/library/ethers.tsx [hunk 2]
```diff
@@ -23,8 +26,36 @@ const modal = createWeb3Modal({
 })
 
 ThemeStore.setModal(modal)
-
+let loaded = false
```
[the remark this answers is not in the document]
Comment (ganchoradkov) [reply apps/laboratory/src/pages/library/ethers.tsx:29]: There seems to be a bug in the app causing components to be created twice or atleast the useEffect is called twice, using state doesn't solve the duplicate call

**Review by chris13524 — COMMENTED** [review]
This needs to be for all pages not just ethers and wagmi

File: apps/laboratory/src/pages/library/ethers.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (enesozturk) [reply apps/laboratory/src/pages/library/ethers.tsx:29]: This is very common thing (not problem) in React world that Strict Mode - which is enabled by default - re-renders the components twice to force you to make sure you don't have unnecessary side effects and your components are pure. Bc when people have different outputs between renders, it's generally sign that they are doing something wrong.

See: https://react.dev/reference/react/StrictMode#fixing-bugs-found-by-double-rendering-in-development

The better approach would be to move this to state and find better approach to don't have issues when re-render. But I still didn't reviewed whole code - just sharing my strict mode knowledge

@ganchoradkov @tomiir

File: apps/laboratory/src/pages/library/ethers.tsx [hunk 3]
```diff
@@ -23,8 +26,36 @@ const modal = createWeb3Modal({
...
+  React.useEffect(() => {
+    if (loaded) {
+      return
+    }
+    loaded = true
+    // @ts-expect-error - walletConnectProviderInitPromise is private
+    modal.walletConnectProviderInitPromise.then(() => {
+      // @ts-expect-error - walletConnectProvider is private
+      const relayer = modal.walletConnectProvider.signer.client.core.relayer
+      relayer.on('relayer_connect', () => {
```
Comment (enesozturk) [inline apps/laboratory/src/pages/library/ethers.tsx:42]: What about removing event listeners on unmount?

Comment (enesozturk) [reply apps/laboratory/src/pages/library/ethers.tsx:42]: Would this refactor help for it?

```jsx
export default function Ethers() {
  const toast = useChakraToast();

  async function initializeWalletConnect(toast) {
    if (loaded) {
      return;
    }
    loaded = true;
    try {
      // @ts-expect-error - walletConnectProviderInitPromise is private
      await modal.walletConnectProviderInitPromise;
  
      return createEventListeners(toast);
    } catch (error) {
      console.error("Error initializing WalletConnect:", error);
    }
  }
  
  function createEventListeners(toast) {
    // @ts-expect-error - walletConnectProvider is private
    const relayer = modal.walletConnectProvider.signer.client.core.relayer;
  
    function handleConnect() {
      toast({
        title: "Connected!",
        description: "wss connection established",
        status: "success",
      });
    }
  
    function handleDisconnect() {
      toast({
        title: "Disconnected!",
        description: "wss connection lost",
        status: "error",
      });
    }
  
    relayer.on("relayer_connect", handleConnect);
    relayer.on("relayer_disconnect", handleDisconnect);
  
    // Return a cleanup function to remove the listeners
    return () => {
      relayer.off("relayer_connect", handleConnect);
      relayer.off("relayer_disconnect", handleDisconnect);
    };
  }

  React.useEffect(() => {
    let listenersCleanup;
    
    initializeWalletConnect(toast).then((cleanup) => {
      listenersCleanup = cleanup;
    });

    return () => {
      if (listenersCleanup) {
        listenersCleanup();
      }
    };
  }, [toast]);

  return (
    <>
      <Web3ModalButtons />
    </>
  );
}
```

Comment (enesozturk) [reply apps/laboratory/src/pages/library/ethers.tsx:42]: @ganchoradkov Do we have an `remove` / `off` function to remove event listener?

**Comment by tomiir** [comment]
Closing and re-building this

**Pull request closed without merging**